servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/KeepAliveManagerTest.java

```diff
@@ -338,4 +396,33 @@ private static final class ScheduledTask {
             this.delayMillis = delayMillis;
         }
     }
+
+    private static final class KeepAliveManagerHandler extends ChannelInboundHandlerAdapter {
+        private KeepAliveManager keepAliveManager;
+
+        void keepAliveManager(KeepAliveManager keepAliveManager) {
+            this.keepAliveManager = requireNonNull(keepAliveManager);
+        }
+
+        @Override
+        public void channelInactive(final ChannelHandlerContext ctx) {
+            keepAliveManager.channelClosed();
+        }
+
+        @Override
+        public void handlerRemoved(final ChannelHandlerContext ctx) {
+            keepAliveManager.channelClosed();
+        }
+
+        @Override
+        public void userEventTriggered(final ChannelHandlerContext ctx, final Object evt) throws Exception {
+            if (evt == ChannelInputShutdownReadComplete.INSTANCE || evt == SslCloseCompletionEvent.SUCCESS) {
+                keepAliveManager.channelInputShutdown();
+            } else if (evt == ChannelOutputShutdownEvent.INSTANCE) {
+                keepAliveManager.channelOutputShutdown();
+            } else {
+                release(evt);
```

Comment:
Should any other user-triggered events cause failure? (i.e. they are unexpected?)

Reply:
I don't think so. My understanding is that user events are fired as best effort to notify a pipeline about something. Other handlers are free to ignore those events if they do not depend on them. 

This handler is a simplification of what we have in `H2ParentConnectionContext.AbstractH2ParentConnection`.